In [65]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

titanic_df = sns.load_dataset('titanic')
titanic_df.head()


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [66]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [67]:
titanic_df['sex'] = titanic_df['sex'].astype('category')
titanic_df['embarked'] = titanic_df['embarked'].astype('category')
titanic_df['who'] = titanic_df['who'].astype('category')
titanic_df['embark_town'] = titanic_df['embark_town'].astype('category')
titanic_df['alive'] = titanic_df['alive'].map({'yes': True, 'no': False})

titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    category
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    category
 8   class        891 non-null    category
 9   who          891 non-null    category
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    category
 13  alive        891 non-null    bool    
 14  alone        891 non-null    bool    
dtypes: bool(3), category(6), float64(2), int64(4)
memory usage: 50.7 KB


In [68]:
titanic_df.describe(include='all').round(4)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
count,891.0000,891.0000,891,714.0000,891.0000,891.0000,891.0000,889,891,891,891,203,889,891,891
unique,NaN,NaN,2,NaN,NaN,NaN,NaN,3,3,3,2,7,3,2,2
top,NaN,NaN,male,NaN,NaN,NaN,NaN,S,Third,man,True,C,Southampton,False,True
freq,NaN,NaN,577,NaN,NaN,NaN,NaN,644,491,537,537,59,644,549,537
mean,0.3838,2.3086,NaN,29.6991,0.5230,0.3816,32.2042,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,0.4866,0.8361,NaN,14.5265,1.1027,0.8061,49.6934,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,0.0000,1.0000,NaN,0.4200,0.0000,0.0000,0.0000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,0.0000,2.0000,NaN,20.1250,0.0000,0.0000,7.9104,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,0.0000,3.0000,NaN,28.0000,0.0000,0.0000,14.4542,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,1.0000,3.0000,NaN,38.0000,1.0000,0.0000,31.0000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


A single column should be created indicating the number of relatives for each passenger on board, instead of:
- number of siblings or spouses on board;
- number of parents or children on board;
Boolean attribute: column 'alone'=True if the passenger was traveling alone (no relatives on board).

In [69]:
titanic_df['relatives'] = titanic_df['sibsp'] + titanic_df['parch']
titanic_df.loc[titanic_df['alone'] == True, 'relatives'] = 0
titanic_df = titanic_df.drop(['sibsp', 'parch','alone'], axis=1)
titanic_df

,survived,pclass,sex,age,fare,embarked,class,who,adult_male,deck,embark_town,alive,relatives
0,0,3,male,22.0,7.2500,S,Third,man,True,NaN,Southampton,False,1
1,1,1,female,38.0,71.2833,C,First,woman,False,C,Cherbourg,True,1
2,1,3,female,26.0,7.9250,S,Third,woman,False,NaN,Southampton,True,0
3,1,1,female,35.0,53.1000,S,First,woman,False,C,Southampton,True,1
4,0,3,male,35.0,8.0500,S,Third,man,True,NaN,Southampton,False,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,13.0000,S,Second,man,True,NaN,Southampton,False,0
887,1,1,female,19.0,30.0000,S,First,woman,False,B,Southampton,True,0
888,0,3,female,NaN,23.4500,S,Third,woman,False,NaN,Southampton,False,3
889,1,1,male,26.0,30.0000,C,First,man,True,C,Cherbourg,True,0


In [70]:
empty_values = titanic_df['relatives'].isna().sum()
print(empty_values)

0


To check how often a certain number of relatives appear in the new column.

In [71]:
titanic_df_frq = titanic_df.groupby('relatives')['relatives'].agg(['count'])

titanic_df_frq

,count
relatives,
0,537
1,161
2,102
3,29
4,15
5,22
6,12
7,6
10,7


Using a loop, replace the number of relatives exceeding the number 5 with the value "above 5".

In [ ]:
titanic_df['relatives_category'] = titanic_df['relatives'].astype(object)

for i in titanic_df.index:
  if titanic_df.loc[i,'relatives'] > 5:
    titanic_df.loc[i, 'relatives_category'] = 'above 5'
#titanic_df.head(20)

titanic_df_relatives = titanic_df.groupby('relatives_category')['relatives_category'].agg(['count'])
titanic_df_relatives.head(20)

,count
relatives_category,
0,537
1,161
2,102
3,29
4,15
5,22
above 5,25


- count the total number of passengers in each of the cities of departure,
- find the number of passengers with more than 5 relatives in each of the cities of departure.

In [73]:
titanic_df_stat = titanic_df.groupby('embark_town', observed=True)['embark_town'].count()
#print(titanic_df_stat)
titanic_df_pas_town = titanic_df[titanic_df['relatives_category'] == 'above 5'].groupby('embark_town', observed=True)['relatives_category'].count()
#print(titanic_df_pas_town)
result = (titanic_df_pas_town / titanic_df_stat * 100).round(2)
result = result.fillna(0)
result.name = 'percentage'
result

embark_town
Cherbourg      0.00
Queenstown     0.00
Southampton    3.88
Name: percentage, dtype: float64

Fill in missing age values ​​with the median.

In [74]:
median_age = titanic_df['age'].median()
#titanic_df['age'].fillna(median_age).info()
titanic_df['age'].fillna(median_age)

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: age, Length: 891, dtype: float64

Create a new column where age is represented by a category instead of a number.

In [ ]:
def get_age_category(age):
  if pd.isna(age):
    return 'unknown'
  elif age < 14:
    return 'child'
  elif age < 35:
    return 'young'
  elif age < 59:
    return 'adult'
  else:
    return 'senior'

titanic_df['age_category'] = titanic_df['age'].apply(get_age_category)
titanic_df

,survived,pclass,sex,age,fare,embarked,class,who,adult_male,deck,embark_town,alive,relatives,relatives_category,age_category
0,0,3,male,22.0,7.2500,S,Third,man,True,NaN,Southampton,False,1,1,young
1,1,1,female,38.0,71.2833,C,First,woman,False,C,Cherbourg,True,1,1,adult
2,1,3,female,26.0,7.9250,S,Third,woman,False,NaN,Southampton,True,0,0,young
3,1,1,female,35.0,53.1000,S,First,woman,False,C,Southampton,True,1,1,adult
4,0,3,male,35.0,8.0500,S,Third,man,True,NaN,Southampton,False,0,0,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,13.0000,S,Second,man,True,NaN,Southampton,False,0,0,young
887,1,1,female,19.0,30.0000,S,First,woman,False,B,Southampton,True,0,0,young
888,0,3,female,NaN,23.4500,S,Third,woman,False,NaN,Southampton,False,3,3,unknown
889,1,1,male,26.0,30.0000,C,First,man,True,C,Cherbourg,True,0,0,young


Check which age group had the highest mortality rate.
Then count the total number of passengers and the number of those who did not survive.
Then calculate the relative rates for each category.

In [ ]:
dead_in_cat = titanic_df[titanic_df['alive'] == False].groupby('age_category')['alive'].count()
total_in_cat = titanic_df.groupby('age_category')['alive'].count()
rate_in_cat = (dead_in_cat / total_in_cat * 100).round(2)

print('\nDead in cat:', dead_in_cat) #total number of those who did not survive for each category
print('\nTotal in cat:', total_in_cat) #total number of passengers for each category
print('\nRate in %:',rate_in_cat)
print('\n', rate_in_cat.idxmax())


Dead in cat: age_category
adult      120
child       29
senior      21
unknown    125
young      254
Name: alive, dtype: int64

Total in cat: age_category
adult      207
child       71
senior      28
unknown    177
young      408
Name: alive, dtype: int64

Rate in %: age_category
adult      57.97
child      40.85
senior     75.00
unknown    70.62
young      62.25
Name: alive, dtype: float64

 senior


Detailed mortality statistics by age category, ticket class, cabin level, and number of relatives.

In [77]:
dead_in_cat_age = titanic_df[titanic_df['alive'] == False].groupby('age_category')['alive'].count()
total_in_cat_age = titanic_df.groupby('age_category')['alive'].count()
rate_in_cat_age = (dead_in_cat_age / total_in_cat_age * 100).round(2)
print('Age category:', rate_in_cat_age)
print('\n', rate_in_cat_age[rate_in_cat_age == rate_in_cat_age.max()])

dead_in_cat_class = titanic_df[titanic_df['alive'] == False].groupby('class', observed=True)['alive'].count()
total_in_cat_class = titanic_df.groupby('class', observed=True)['alive'].count()
rate_in_cat_class = (dead_in_cat_class / total_in_cat_class * 100).round(2)
print('\nClass category:', rate_in_cat_class)
print('\n', rate_in_cat_class[rate_in_cat_class == rate_in_cat_class.max()])

dead_in_cat_deck = titanic_df[titanic_df['alive'] == False].groupby('deck', observed=True)['alive'].count()
total_in_cat_deck = titanic_df.groupby('deck', observed=True)['alive'].count()
rate_in_cat_deck = (dead_in_cat_deck / total_in_cat_deck * 100).round(2)
print('\nDeck category:', rate_in_cat_deck)
print('\n', rate_in_cat_deck[rate_in_cat_deck == rate_in_cat_deck.max()])

dead_in_cat_relatives = titanic_df[titanic_df['alive'] == False].groupby('relatives', observed=True)['alive'].count()
total_in_cat_relatives = titanic_df.groupby('relatives', observed=True)['alive'].count()
rate_in_cat_relatives = (dead_in_cat_relatives / total_in_cat_relatives * 100).round(2)
print('\nRelatives category:', rate_in_cat_relatives)
print('\n', rate_in_cat_relatives[rate_in_cat_relatives == rate_in_cat_relatives.max()])

Age category: age_category
adult      57.97
child      40.85
senior     75.00
unknown    70.62
young      62.25
Name: alive, dtype: float64

 age_category
senior    75.0
Name: alive, dtype: float64

Class category: class
First     37.04
Second    52.72
Third     75.76
Name: alive, dtype: float64

 class
Third    75.76
Name: alive, dtype: float64

Deck category: deck
A    53.33
B    25.53
C    40.68
D    24.24
E    25.00
F    38.46
G    50.00
Name: alive, dtype: float64

 deck
A    53.33
Name: alive, dtype: float64

Relatives category: relatives
0      69.65
1      44.72
2      42.16
3      27.59
4      80.00
5      86.36
6      66.67
7     100.00
10    100.00
Name: alive, dtype: float64

 relatives
7     100.0
10    100.0
Name: alive, dtype: float64


The highest mortality rate was observed among passengers in the senior age group, with a mortality rate of 75.00%.

Among ticket classes, third-class passengers experienced the highest mortality rate at 75.76%.

Passengers traveling with 7 or 10 relatives had a 100% mortality rate, meaning no one in these groups survived.

Regarding cabin levels, deck G was associated with the highest mortality rate at 50.00%, making it the riskiest cabin deck.

Overall, ticket class and age appear to have had the greatest impact on passenger mortality.